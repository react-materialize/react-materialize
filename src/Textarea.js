import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import idgen from './idgen';
import constants from './constants';

const Textarea = ({
  children,
  className,
  s,
  m,
  l,
  xl,
  noLayout,
  placeholder,
  icon,
  label,
  success,
  error,
  iconClassName,
  onChange,
  validate,
  defaultValue,
  value,
  id,
  ...rest
}) => {
  const textareaRef = useRef(null);
  const dataLength = rest['data-length'];

  useEffect(() => {
    if (dataLength) {
      M.CharacterCounter.init(textareaRef.current);
    }
  }, [dataLength, children]);

  useEffect(() => {
    M.textareaAutoResize(textareaRef.current);
  }, [children]);

  const sizes = { s, m, l, xl };

  let responsiveClasses;
  if (!noLayout) {
    responsiveClasses = { col: true };
    constants.SIZES.forEach(size => {
      responsiveClasses[size + sizes[size]] = sizes[size];
    });
  }

  return (
    <div className={cx('input-field', responsiveClasses)}>
      {Boolean(icon) && React.cloneElement(icon, { className: 'prefix' })}
      <textarea
        placeholder={placeholder}
        id={id}
        value={value}
        defaultValue={defaultValue}
        ref={textareaRef}
        onChange={onChange}
        {...rest}
        className={cx(
          'materialize-textarea',
          { validate },
          { invalid: error },
          className
        )}
      />
      {Boolean(label) && (
        <label
          className={cx({
            active: value || placeholder,
            'label-icon': typeof label !== 'string'
          })}
          data-success={success}
          data-error={error}
          htmlFor={id}
        >
          {label}
        </label>
      )}
      {[error || success] && (
        <span
          className="helper-text"
          data-error={error}
          data-success={success}
        />
      )}
      {children}
    </div>
  );
};

Textarea.propTypes = {
  children: PropTypes.node,
  /**
   * Strip away all layout classes such as col and sX
   */
  noLayout: PropTypes.bool,
  /**
   * Responsive size for Mobile Devices
   */
  s: PropTypes.number,
  /**
   * Responsive size for Tablet Devices
   */
  m: PropTypes.number,
  /**
   * Responsive size for Desktop Devices
   */
  l: PropTypes.number,
  /**
   *  Responsive size for Large Desktop Devices
   */
  xl: PropTypes.number,
  /**
   * disabled input
   */
  disabled: PropTypes.bool,
  /**
   * Placeholder string
   */
  placeholder: PropTypes.string,
  /**
   * override id
   * @default idgen()
   */
  id: PropTypes.string,
  /**
   * render icon next to input
   */
  icon: PropTypes.node,
  /**
   * icon classname
   */
  iconClassName: PropTypes.string,
  /**
   * textarea label
   */
  label: PropTypes.string,
  /**
   * Input initial value
   */
  defaultValue: PropTypes.string,
  /**
   * predefined value
   */
  value: PropTypes.string,
  /**
   * Add validate class to input
   */
  validate: PropTypes.bool,
  /**
   * Custom success message
   */
  success: PropTypes.string,
  /**
   * Custom error message
   */
  error: PropTypes.string,
  /**
   * textarea classname
   */
  className: PropTypes.string,
  /**
   * onChange callback
   */
  onChange: PropTypes.func
};

Textarea.defaultProps = {
  get id() {
    return `Textarea-${idgen()}`;
  }
};

export default Textarea;
