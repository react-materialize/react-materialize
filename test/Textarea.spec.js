import React from 'react';
// import { shallow, mount } from 'enzyme';
import Textarea from '../src/Textarea';

describe.skip('<Textarea />', () => {
  let wrapper;

  test('Textarea', () => {
    wrapper = shallow(<Textarea />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Handles layout classes', () => {
    wrapper = shallow(<Textarea s={4} m={6} />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeTruthy();
    expect(wrapper.hasClass('s4')).toBeTruthy();
    expect(wrapper.hasClass('m6')).toBeTruthy();
  });

  test('Handles removing layout classes', () => {
    wrapper = shallow(<Textarea s={4} m={6} noLayout />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeFalsy();
    expect(wrapper.hasClass('s4')).toBeFalsy();
    expect(wrapper.hasClass('m6')).toBeFalsy();
  });

  test('Value', () => {
    wrapper = shallow(<Textarea value="I was here" />);
    expect(wrapper.find('textarea').prop('value')).toEqual('I was here');
    expect(wrapper).toMatchSnapshot();
  });

  test('defaultValue renders as value', () => {
    wrapper = mount(<Textarea defaultValue="I was here" />);
    expect(wrapper.find('textarea').prop('defaultValue')).toEqual('I was here');
    expect(wrapper.find('textarea').instance().value).toEqual('I was here');
  });

  test('Disabled', () => {
    wrapper = shallow(<Textarea value="I was here" disabled />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Placeholder', () => {
    wrapper = shallow(<Textarea placeholder="I was here" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Label', () => {
    wrapper = shallow(<Textarea label="Write something here..." />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Icon', () => {
    wrapper = shallow(<Textarea icon="mode_edit" iconClassName="red" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('with custom className', () => {
    wrapper = shallow(<Textarea className="custom-class" />);
    const textareaElement = wrapper.find('textarea');
    expect(textareaElement.hasClass('materialize-textarea')).toBeTruthy();
    expect(textareaElement.hasClass('custom-class')).toBeTruthy();
  });

  test('onChange is fired', () => {
    const onChangeMock = jest.fn();
    wrapper = shallow(<Textarea onChange={onChangeMock} />);
    const textarea = wrapper.find('textarea');
    textarea.simulate('change');
    expect(onChangeMock).toHaveBeenCalled();
  });

  test('Displays error messages', () => {
    wrapper = shallow(
      <Textarea label="Input with Error" error="this is an error" />
    );
    expect(wrapper.find('textarea').hasClass('invalid')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
