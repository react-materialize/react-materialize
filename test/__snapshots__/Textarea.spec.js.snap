// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea /> Disabled 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    disabled={true}
    id="Textarea-mockId"
    value="I was here"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Displays error messages 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea invalid"
    id="Textarea-mockId"
  />
  <label
    className=""
    data-error="this is an error"
    htmlFor="Textarea-mockId"
  >
    Input with Error
  </label>
  <span
    className="helper-text"
    data-error="this is an error"
  />
</div>
`;

exports[`<Textarea /> Icon 1`] = `
<div
  className="input-field col"
>
  <Component
    className="prefix"
  />
  <textarea
    className="materialize-textarea"
    id="Textarea-mockId"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Label 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="Textarea-mockId"
  />
  <label
    className=""
    htmlFor="Textarea-mockId"
  >
    Write something here...
  </label>
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Placeholder 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="Textarea-mockId"
    placeholder="I was here"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Textarea 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="Textarea-mockId"
  />
  <span
    className="helper-text"
  />
</div>
`;

exports[`<Textarea /> Value 1`] = `
<div
  className="input-field col"
>
  <textarea
    className="materialize-textarea"
    id="Textarea-mockId"
    value="I was here"
  />
  <span
    className="helper-text"
  />
</div>
`;
